pkgs/top-level/python-packages.nix

```
@@ -4019,7 +4019,16 @@ in {
 
   jax = callPackage ../development/python-modules/jax { };
 
-  jaxlib = callPackage ../development/python-modules/jaxlib { };
+  jaxlib-bin = callPackage ../development/python-modules/jaxlib/bin.nix { };
+
+  jaxlib-build = callPackage ../development/python-modules/jaxlib {
+    cudaSupport = pkgs.config.cudaSupport or false;
```

```suggestion
    cudaSupport = pkgs.config.cudaSupport;
```
since `x or false == x`

---

This is copy-paste from `tensorflow` build. Looks like it has to be this way due to missing `cudaSupport` attribute on some platforms (like `*-darwin`), see the failure @ https://github.com/NixOS/nixpkgs/runs/4626725141?check_suite_focus=true if I try to use just `pkgs.config.cudaSupport`.

---

ah ok can we add a comment explaining this?

---

It's not always true that `x or false == x`. If x is `undefined`/`null` then the result of this is `false`.

This idiom is used to set the default value in case the variable is not set and one can see it plenty of times in dynamic programming languages codebases such as Javascript (`|| <default_val>`) and Python (`or <default_val>`).

---

In this case I believe the original is correct.
```
    cudaSupport = pkgs.config.cudaSupport or false;
```

The `pkgs.config` set is not expected to contain configuration for all keys. The failure in the checks is trying to read from `pkgs.config`, which is likely empty or close to empty.

> It's not always true that `x or false == x`. If x is `undefined`/`null` then the result of this is `false`.

This is true in some languages, but nix has two distinct operators. `||`, which only works on booleans (so, `null || false` is an error), and the defaulting `or`, which is for providing defaults for absent keys in a set and only works in attribute lookups. There’s a mention of it in the [documentation for Set](https://nixos.org/manual/nix/stable/expressions/language-values.html#sets), but it’s not super clear.

```
nix-repl> null || false
error: value is null while a Boolean was expected, at (string):1:1

nix-repl> {}.a or "default"
"default"

nix-repl> { a = null; }.a or "default"
null

nix-repl> { a = false; }.a or "default"
false

nix-repl> true or false
error: undefined variable 'or' at (string):1:1
```

---

Yes, reverted to the original version (with `or false`) and added a comment.

---

Ah ok, thanks for clarifying! I stand corrected...